libs/constellation/src/constellation.cpp

```diff
@@ -338,69 +360,225 @@ Constellation::Constellation(CertificatePtr const &certificate, Config config)
+                                                      cfg_.log2_num_lanes);
+
+  return true;
+}
+
+bool Constellation::OnRestorePreviousData(ledger::GenesisFileCreator::ConsensusParameters &params)
+{
+  FETCH_LOG_INFO(LOGGING_NAME, "OnRestorePreviousData()");
+
+  GenesisFileCreator::Result genesis_status = GenesisFileCreator::Result::FAILURE;
+
+  // attempt to do of the following things
+  // - perform initial start up of the system state
+  // - recover from previous genesis init
+  {
+    std::string const genesis_file_path =
+        cfg_.genesis_file_location.empty() ? GENESIS_FILENAME : cfg_.genesis_file_location;
+
+    GenesisFileCreator creator(*storage_, external_identity_, cfg_.db_prefix);
+
+    genesis_status = creator.LoadFile(genesis_file_path, cfg_.proof_of_stake, params);
+
+    if (genesis_status == GenesisFileCreator::Result::FAILURE)
+    {
+      FETCH_LOG_ERROR(LOGGING_NAME, "Failed to restore or generate Genesis configuration");
+      return false;
+    }
+  }
+
+  // create the DAG
+  dag_ = GenerateDAG(cfg_, "dag_db_", true, external_identity_);
+
+  // create the chain
+  chain_ = std::make_unique<MainChain>(ledger::MainChain::Mode::LOAD_PERSISTENT_DB);
+
+  if (!GenesisSanityChecks(genesis_status))
+  {
+    return false;
+  }
+
+  if (!CheckStateIntegrity())
+  {
+    return false;
+  }
+
+  auto const heaviest_block = chain_->GetHeaviestBlock();
+  FETCH_LOG_INFO(LOGGING_NAME, "Head of chain: #", heaviest_block->block_number, " 0x",
+                 heaviest_block->hash.ToHex(), " Merkle: 0x", heaviest_block->merkle_hash.ToHex());
+
+  return true;
+}
+
+bool Constellation::OnBringUpExternalNetwork(
+    ledger::GenesisFileCreator::ConsensusParameters &params, UriSet const &initial_peers)
+{
+  FETCH_LOG_INFO(LOGGING_NAME, "OnBringUpExternalNetwork()");
+
+  muddle_ = muddle::CreateMuddle("IHUB", external_identity_, network_manager_,
+                                 cfg_.manifest.FindExternalAddress(ServiceIdentifier::Type::CORE));
+
+  shard_management_ = std::make_shared<ShardManagementService>(cfg_.manifest, *lane_control_,
+                                                               *muddle_, cfg_.log2_num_lanes);
+
+  // setup the consensus infrastructure
+  beacon_network_ = CreateBeaconNetwork(cfg_, external_identity_, network_manager_);
+  beacon_setup_ =
+      CreateBeaconSetupService(cfg_, *beacon_network_, *shard_management_, external_identity_);
+  beacon_ = CreateBeaconService(cfg_, *beacon_network_, external_identity_, beacon_setup_);
+  stake_  = CreateStakeManager(cfg_);
+
+  consensus_ = CreateConsensus(cfg_, stake_, beacon_setup_, beacon_, *chain_, *storage_,
+                               external_identity_->identity());
+
+  if (cfg_.proof_of_stake)
+  {
+    consensus_->SetDefaultStartTime(params.start_time);
+    consensus_->SetMaxCabinetSize(params.cabinet_size);
+
+    if (params.snapshot)
+    {
+      consensus_->Reset(*params.snapshot);
```

Comment:
I will probably have to change this in my wealth PR - the flow should be basically the same for POS